zerver/tests/test_auth_backends.py

```diff
@@ -2451,6 +2478,36 @@ def test_get_external_method_buttons(self) -> None:
             self.assert_in_success_response([string.format("register") for string in expected_button_id_strings],
                                             result)
 
+    def test_get_external_method_dicts_multiple_saml_idps(self) -> None:
+        idps_dict = copy.deepcopy(settings.SOCIAL_AUTH_SAML_ENABLED_IDPS)
+        # Create another IdP config, by copying the original one and changing some details.idps_dict['test_idp'])
+        idps_dict['test_idp2'] = copy.deepcopy(idps_dict['test_idp'])
+        idps_dict['test_idp2']['url'] = 'https://idp2.example.com/idp/profile/SAML2/Redirect/SSO'
+        idps_dict['test_idp2']['display_name'] = 'Second Test IdP'
+        idps_dict['test_idp2']['limit_to_subdomains'] = ['zephyr', ]
+        with self.settings(
+            SOCIAL_AUTH_SAML_ENABLED_IDPS=idps_dict,
+            AUTHENTICATION_BACKENDS=('zproject.backends.EmailAuthBackend',
+                                     'zproject.backends.GitHubAuthBackend',
+                                     'zproject.backends.SAMLAuthBackend')
+        ):
+            # Calling get_external_method_dicts without a realm returns all methods configured on the server:
+            external_auth_methods = get_external_method_dicts()
+            self.assert_length(external_auth_methods, 3)  # 2 IdP + a dict for github auth
+            self.assertEqual(set([external_auth_methods[0]['name'], external_auth_methods[1]['name']]),
+                             set(['saml:test_idp', 'saml:test_idp2']))
```

Comment:
Hmm, this is a bit awkward.  I don't think we want to return the list of SAML backends on the server here, since they will likely have names like `customername1`, etc.  I don't think we actually use this endpoint on the root domain for much; the most we're likely to use it for is authentication for /new.  So probably it's reasonable to just filter `realm=None`.

Reply:
Agreed! In order to be consistent across all the backends that we have I added a commit to just return nothing if queried without a realm. (If we changed this only for SAML IdPs, this behavior would be inconsistent because we'd still be returning Google/Github/etc. dicts when querying with no realm).